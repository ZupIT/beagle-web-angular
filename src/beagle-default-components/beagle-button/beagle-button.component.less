/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

beagle-button {
    .default-theme {
        border: 1px solid currentColor;
        line-height: 34px;
        color: inherit;
        background: transparent;
        border-color:rgba(36, 42, 49, 0.7);
        box-sizing: border-box;
        position: relative;
        cursor: pointer;
        outline: none;
        display: inline-block;
        white-space: nowrap;
        text-decoration: none;
        vertical-align: baseline;
        text-align: center;
        margin: 0;
        min-width: 70px;
        line-height: 36px;
        padding: 0 16px;
        border-radius: 5px;
        overflow: visible;
        &:hover {
          background-color:  rgba(36, 42, 49, 0.1);
        }
    }
}