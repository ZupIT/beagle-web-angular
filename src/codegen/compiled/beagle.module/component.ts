/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Type } from '@angular/core'
import {
  kebabToCamelCase,
  removeExtraIndentation,
  replaceToUnderline,
} from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  const componentName = replaceToUnderline(name)

  if (angularVersion >= 8) {
    return `'${componentName}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${componentName}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(
    components: Record<string, Type<any>>,
    angularVersion: number,
  ) {
  const queries = createQueries(components, angularVersion)

  // todo: legacy code. Remove the input "loadParams" with v2.0.
  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams', 'route', 'networkOptions', 'controllerId'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
