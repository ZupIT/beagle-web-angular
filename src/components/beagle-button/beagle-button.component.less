/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

@import "../assets/default-style.less";

beagle-button {
  background-color: inherit !important;
  padding: 0 !important;
  display: inline-block !important;
  
  button {
    cursor: pointer;
    border: 1px solid @swampLight;
    color: inherit;
    background: transparent;
    box-sizing: border-box;
    position: relative;
    outline: none;
    display: inline-block;
    white-space: nowrap;
    text-decoration: none;
    vertical-align: baseline;
    text-align: center;
    margin: 0;
    min-width: 70px;
    width: 100%;
    line-height: 34px;
    padding: 0 16px;
    border-radius: 5px;
    overflow: visible;

    &:hover {
      background-color: @swampTransparent;
    }
  }
}