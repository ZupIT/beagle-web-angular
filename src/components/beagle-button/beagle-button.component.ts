/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/
import { AfterViewInit, Component, Input, ViewEncapsulation } from '@angular/core'
import { ClickEvent, IdentifiableBeagleUIElement } from '@zup-it/beagle-web'
import { Properties as CSSProperties } from 'csstype'

import { BeagleComponent } from '../../runtime/BeagleComponent'
import { BeagleButtonInterface, StylesNotToInherit } from '../schemas/button'

@Component({
  selector: 'beagle-button',
  templateUrl: './beagle-button.component.html',
  styleUrls: ['./beagle-button.component.less'],
  encapsulation: ViewEncapsulation.None,
})
export class BeagleButtonComponent extends BeagleComponent
  implements BeagleButtonInterface, AfterViewInit {

  @Input() text: string
  @Input() styleId?: string
  @Input() onPress?: () => void
  @Input() style?: CSSProperties
  @Input() clickAnalyticsEvent?: ClickEvent
  @Input() enabled?: boolean
  public usefulStyle: Record<string, any> = {}
  public type = 'button'
  public disabled: boolean

  ngOnInit() {
    if (this.style) {
      this.usefulStyle = Object.keys(this.style).reduce((styleObject, prop) => {
        if (!StylesNotToInherit.includes(prop)) {
          styleObject[prop] = this.style && this.style[prop]
        }
        return styleObject
      }, {})
    }
  }

  ngOnChanges() {
    this.disabled = this.enabled === undefined ? false : !this.enabled
  }

  ngAfterViewInit() {
    if (this.getViewContentManager instanceof Function) {
      const element = this.getViewContentManager().getElement()
      this.type = this.isSubmitButton(element)
    }
  }

  handleClick() {
    this.onPress && this.type === 'button' && this.onPress()

    if (this.getViewContentManager instanceof Function) {
        const beagleService = this.getViewContentManager().getView().getBeagleService()
        const analytics = beagleService ? beagleService.analytics : null

        if (this.clickAnalyticsEvent && analytics) {
          analytics.trackEventOnClick(this.clickAnalyticsEvent)
        }
    }    
  }

  private isSubmitButton(element: IdentifiableBeagleUIElement<any> | null) {
    return element &&
      element.onPress &&
      element.onPress._beagleAction_ === 'beagle:submitForm' ? 'submit' : 'button'
  }
}
