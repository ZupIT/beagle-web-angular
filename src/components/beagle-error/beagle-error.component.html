<!-- 
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License. 
-->

<beagle-text text="Sorry!" textColor="red"></beagle-text>
<beagle-text text="An unexpected error happened while loading your page."></beagle-text>
<beagle-button [onPress]="retry" class="retry-button" text="Retry"></beagle-button>

<ng-container *ngIf="errors?.length > 0">
  <beagle-button [onPress]="toggleDetails" class="show-details-button" text="{{ showingDetails ? 'Hide' : 'Show' }} details {{ showingDetails ? '▲' : '▼' }}"></beagle-button>
  <section class="errors-detailed-container" [ngClass]="{ 'show': showingDetails }">
    <p *ngFor="let error of errors" class="error-message">
      <span class="error-indicator">►</span> ERROR: {{ error.message }}
    </p>
  </section>
</ng-container>
