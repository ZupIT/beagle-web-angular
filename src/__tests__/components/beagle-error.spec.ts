/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { DebugElement } from '@angular/core'
import { By } from '@angular/platform-browser'
import { TestBed, async, ComponentFixture } from '@angular/core/testing'
import { BeagleErrorComponent } from '../../components/beagle-error/beagle-error.component'
import { BeagleButtonComponent } from '../../components/beagle-button/beagle-button.component'
import { BeagleTextComponent } from '../../components/beagle-text/beagle-text.component'

let component: BeagleErrorComponent
let fixture: ComponentFixture<BeagleErrorComponent>
let compiledNative: HTMLElement
let compiledDebug: DebugElement

describe('BeagleErrorComponent - empty', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        BeagleTextComponent,
        BeagleButtonComponent,
        BeagleErrorComponent,
      ],
    }).compileComponents()
  }))

  beforeEach(() => {
    fixture = TestBed.createComponent(BeagleErrorComponent)
    compiledNative = fixture.debugElement.nativeElement
    compiledDebug = fixture.debugElement

    component = fixture.componentInstance
    component.retry = jest.fn()
    component.errors = []
    component.showingDetails = false

    fixture.detectChanges()
  })

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  it('Should not render details when no error is provided', () => {
    const showDetailsButton = compiledNative.querySelectorAll('.show-details-button')
    const errorsDetailedContainer = compiledNative.querySelectorAll('.errors-detailed-container')

    expect(showDetailsButton.length).toBe(0)
    expect(errorsDetailedContainer.length).toBe(0)
  })

  it('Should call retry function on click', () => {
    const button = compiledDebug.query(By.css('.retry-button'))
      .componentInstance as BeagleButtonComponent
    button.handleClick()
    expect(component.retry).toHaveBeenCalled()
  })
})

describe('BeagleErrorComponent - with errors', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        BeagleTextComponent,
        BeagleButtonComponent,
        BeagleErrorComponent,
      ],
    }).compileComponents()
  }))

  beforeEach(() => {
    fixture = TestBed.createComponent(BeagleErrorComponent)
    compiledNative = fixture.debugElement.nativeElement
    compiledDebug = fixture.debugElement

    component = fixture.componentInstance
    component.retry = jest.fn()
    component.errors = [{ message: 'First Message' }, { message: 'Second Message' }]
    component.showingDetails = false

    fixture.detectChanges()
  })

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  it('Should call retry function on click', () => {
    const button = compiledDebug.query(By.css('.retry-button'))
      .componentInstance as BeagleButtonComponent
    button.handleClick()
    expect(component.retry).toHaveBeenCalled()
  })

  it('Should render details when error list is provided', () => {
    const showDetailsButton = compiledNative.querySelectorAll('.show-details-button')
    const errorsDetailedContainer = compiledNative.querySelectorAll('.errors-detailed-container')

    expect(showDetailsButton.length).toBe(1)
    expect(errorsDetailedContainer.length).toBe(1)

    const errorMessages = errorsDetailedContainer[0].querySelectorAll('p')

    expect(errorMessages.length).toBe(2)
  })

  it('Should not add the class when the show more button was not clicked', () => {
    const errorsDetailedContainer = compiledNative.querySelector('.errors-detailed-container')
    const showContainer = errorsDetailedContainer?.querySelectorAll('.show')

    expect(showContainer?.length).toBe(0)
  })

  it('Should add the class when the show more button was clicked', () => {
    const button = compiledDebug.query(By.css('.show-details-button'))
      .componentInstance as BeagleButtonComponent
    button.handleClick()
    fixture.detectChanges()

    const errorsDetailedContainer = compiledNative.querySelector('.errors-detailed-container')

    expect(component.showingDetails).toBeTruthy()
    expect(errorsDetailedContainer?.classList.contains('show')).toBeTruthy()

    button.handleClick()
    fixture.detectChanges()

    expect(component.showingDetails).toBeFalsy()
    expect(errorsDetailedContainer?.classList.contains('show')).toBeFalsy()
  })
})
