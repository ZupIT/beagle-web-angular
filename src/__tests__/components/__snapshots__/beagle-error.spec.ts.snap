// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleErrorComponent - empty should match snapshot 1`] = `
<beagle-error
  errors={[Function Array]}
  retry={[Function Function]}
  showingDetails="false"
  toggleDetails={[Function Function]}
>
  <beagle-text
    text="Sorry!"
    textcolor="red"
  >
    <p
      class="beagle-text"
      style="color: red; text-align: inherit;"
    >
       Sorry!
  
    </p>
  </beagle-text><beagle-text
    text="An unexpected error happened while loading your page."
  >
    <p
      class="beagle-text"
      style="text-align: inherit;"
    >
       An unexpected error happened while loading your page.
  
    </p>
  </beagle-text><beagle-button
    class="retry-button"
    text="Retry"
  >
    <button
      type="button"
    >
       Retry
  
    </button>
  </beagle-button>
</beagle-error>
`;

exports[`BeagleErrorComponent - with errors should match snapshot 1`] = `
<beagle-error
  errors={[Function Array]}
  retry={[Function Function]}
  showingDetails="false"
  toggleDetails={[Function Function]}
>
  <beagle-text
    text="Sorry!"
    textcolor="red"
  >
    <p
      class="beagle-text"
      style="color: red; text-align: inherit;"
    >
       Sorry!
  
    </p>
  </beagle-text><beagle-text
    text="An unexpected error happened while loading your page."
  >
    <p
      class="beagle-text"
      style="text-align: inherit;"
    >
       An unexpected error happened while loading your page.
  
    </p>
  </beagle-text><beagle-button
    class="retry-button"
    text="Retry"
  >
    <button
      type="button"
    >
       Retry
  
    </button>
  </beagle-button><beagle-button
    class="show-details-button"
  >
    <button
      type="button"
    >
       Show details ▼
  
    </button>
  </beagle-button><section
    class="errors-detailed-container"
  >
    
    <p
      class="error-message"
    >
      <span
        class="error-indicator"
      >
        ►
      </span>
       ERROR: First Message 
    </p>
    <p
      class="error-message"
    >
      <span
        class="error-indicator"
      >
        ►
      </span>
       ERROR: Second Message 
    </p>
  </section>
</beagle-error>
`;
